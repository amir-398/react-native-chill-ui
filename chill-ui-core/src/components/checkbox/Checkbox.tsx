import { cn } from '@utils';
import { useState } from 'react';
import { Pressable } from 'react-native';

import { Box } from '../box';
import { Icon } from '../icon';
import { String } from '../string';
import { CheckboxProps } from '../../types/checkbox.types';
import { checkboxSizeVariants, checkboxVariants } from './styleVariants';

/**
 * Checkbox component with support for different variants, custom styling, and accessibility features.
 * Provides a flexible and accessible checkbox implementation with support for controlled and uncontrolled patterns.
 *
 * @example
 * ```tsx
 * // Basic usage
 * const [isChecked, setIsChecked] = useState(false);
 *
 * <Checkbox
 *   label="Accept terms and conditions"
 *   isChecked={isChecked}
 *   onChange={setIsChecked}
 * />
 *
 * // With different variants
 * <Checkbox
 *   label="Circle Checkbox"
 *   variant="circle"
 *   isChecked={isChecked}
 *   onChange={setIsChecked}
 * />
 *
 * // With custom colors
 * <Checkbox
 *   label="Custom Colors"
 *   checkedColor="#FF0000"
 *   uncheckedColor="#CCCCCC"
 *   iconColor="#FFFFFF"
 *   isChecked={isChecked}
 *   onChange={setIsChecked}
 * />
 *
 * // With custom icon
 * <Checkbox
 *   label="Custom Icon"
 *   iconName="star-solid"
 *   iconColor="#FFD700"
 *   isChecked={isChecked}
 *   onChange={setIsChecked}
 * />
 *
 * // With custom size
 * <Checkbox
 *   label="Custom Size"
 *   checkboxSize={30}
 *   isChecked={isChecked}
 *   onChange={setIsChecked}
 * />
 * ```
 *
 * @param checkboxSize - Custom size for the checkbox
 * @param checkedClassName - Additional CSS classes when checked
 * @param checkedColor - Background color when checked
 * @param className - Additional CSS classes for the checkbox container
 * @param customIcon - Custom icon component to use instead of default
 * @param iconColor - Color of the check icon (default: 'white')
 * @param iconName - Name of the icon to use (default: 'check-solid')
 * @param isChecked - Whether the checkbox is checked (default: false)
 * @param isDisabled - Whether the checkbox is disabled (default: false)
 * @param label - Label text for the checkbox
 * @param labelClassName - Additional CSS classes for the label
 * @param onChange - Callback when checkbox state changes
 * @param size - Size variant of the checkbox (default: 'sm')
 * @param uncheckedColor - Background color when unchecked
 * @param variant - Visual variant of the checkbox (default: 'square')
 */
export default function Checkbox({
  checkboxSize,
  checkedClassName,
  checkedColor,
  className,
  customIcon,
  iconColor = 'white',
  iconName = 'check-solid',
  isChecked = false,
  isDisabled = false,
  label,
  labelClassName,
  onChange,
  size = 'sm',
  uncheckedColor,
  variant = 'square',
  ...rest
}: CheckboxProps) {
  const [isDefaultChecked, setIsDefaultChecked] = useState(isChecked);

  const handlePress = () => {
    if (!isDisabled && onChange) {
      onChange(!isDefaultChecked);
      setIsDefaultChecked(!isDefaultChecked);
    }
  };

  return (
    <Pressable onPress={handlePress} disabled={isDisabled} className="flex-row items-center" {...rest}>
      <Box
        className={cn(
          'items-center justify-center border',
          // @ts-ignore
          { [checkboxSizeVariants({ size })]: customIcon, 'opacity-50': isDisabled, 'p-0.5': !customIcon },
          isDefaultChecked && (checkedClassName ?? 'border-primary bg-primary'),
          !isDefaultChecked && 'border-gray-300 bg-white',
          checkboxVariants({ variant }),
          className,
        )}
        style={{
          ...(checkboxSize && { height: checkboxSize, width: checkboxSize }),
          ...(checkedColor && (isDefaultChecked ?? isChecked) && { backgroundColor: checkedColor }),
          ...(uncheckedColor && !isDefaultChecked && !isChecked && { backgroundColor: uncheckedColor }),
        }}
      >
        {isDefaultChecked && customIcon && customIcon}
        {!customIcon && <Icon name={iconName} size={size} color={isDefaultChecked ? iconColor : 'transparent'} />}
      </Box>
      {label && (
        // @ts-ignore
        <String className={cn({ 'opacity-30': isDisabled }, 'ml-2 text-black', labelClassName)} size={size}>
          {label}
        </String>
      )}
    </Pressable>
  );
}
