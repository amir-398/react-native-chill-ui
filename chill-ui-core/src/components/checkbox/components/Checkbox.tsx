import type { CheckboxPropsTw } from '@types';

import { TIcons } from '@constants';
import { Box } from '@components/box';
import { Icon } from '@components/icon';
import { Pressable } from 'react-native';
import { String } from '@components/string';
import { PropsWithChildren, useMemo, useState } from 'react';
import { classNameHandler, classNamePropsHandler, cn, isUndefined, styleHandler } from '@utils';

import { checkboxDefaultProps } from '../utils/defaultProps';
import { checkboxSv, styles } from '../styles/Checkbox.ss.styles';
import { checkboxTv, twStyles } from '../styles/Checkbox.tw.styles';

/**
 * Checkbox component with support for different variants, custom styling, and accessibility features.
 * Provides a flexible and accessible checkbox implementation with support for controlled and uncontrolled patterns.
 * Automatically detects NativeWind availability and falls back to StyleSheet if needed.
 *
 * @example
 * ```tsx
 * <Checkbox
 *   label="Accept terms"
 *   isChecked={checked}
 *   onCheckedChange={setChecked}
 *   variant="circle"
 *   size="md"
 *   isLabelPressable
 * />
 * ```
 *
 * @param checkedClassName - (only NativeWind) Additional CSS classes when checked
 * @param checkedColor - Background color when checked
 * @param checkedStyle - Additional style when checked
 * @param className - (only NativeWind) Additional CSS classes for the checkbox container
 * @param iconProps - Props for the icon component
 * @param isChecked - Whether the checkbox is checked (default: false)
 * @param isDisabled - Whether the checkbox is disabled (default: false)
 * @param isLabelPressable - Whether the label is pressable (default: false)
 * @param label - Label text for the checkbox
 * @param labelProps - Props for the label component
 * @param onCheckedChange - Callback when checkbox state changes
 * @param size - Size variant of the checkbox (default: 'md')
 * @param style - Style object for the checkbox container
 * @param uncheckedColor - Background color when unchecked
 * @param variant - Visual variant of the checkbox (default: 'square')
 */
export default function Checkbox(props: PropsWithChildren<CheckboxPropsTw>) {
  classNamePropsHandler(props, 'Checkbox');

  const {
    checkedClassName,
    checkedColor,
    checkedStyle,
    children,
    className,
    iconProps,
    isChecked: controlledChecked,
    isDisabled = checkboxDefaultProps.isDisabled,
    isLabelPressable = checkboxDefaultProps.isLabelPressable,
    label,
    labelProps,
    onCheckedChange,
    size = checkboxDefaultProps.size,
    style,
    uncheckedColor,
    variant = checkboxDefaultProps.variant,
  } = props;

  const [internalChecked, setInternalChecked] = useState(checkboxDefaultProps.isChecked);

  const checked = isUndefined(controlledChecked) ? internalChecked : controlledChecked;

  const handlePress = () => {
    if (isDisabled) return;

    const newValue = !checked;

    if (isUndefined(controlledChecked)) {
      setInternalChecked(newValue);
    }

    onCheckedChange?.(newValue);
  };

  const iconSize = useMemo(() => {
    const sizeMap: Record<string, string> = {
      '2xl': 'lg',
      lg: 'sm',
      md: 'xs',
      sm: '2xs',
      xl: 'md',
      xs: '3xs',
    };

    return sizeMap[size as keyof typeof sizeMap] || 'xs';
  }, [size]);

  const contentCheckBox = (
    <Box
      {...classNameHandler(
        cn(
          twStyles.contentCheckBox,
          checkboxTv({ isChecked: checked, isDisabled, size, variant }),
          checked && checkedClassName,
          className,
        ),
      )}
      {...styleHandler({
        defaultStyle: [checkboxSv({ isChecked: checked, isDisabled, size, variant })],
        style: [
          {
            ...(checkedColor && checked && { backgroundColor: checkedColor }),
            ...(uncheckedColor && !checked && { backgroundColor: uncheckedColor }),
          },
          checked && checkedStyle,
          style,
        ],
      })}
    >
      {checked &&
        (children ?? (
          <Icon
            name={checkboxDefaultProps.iconName as keyof TIcons}
            color={checkboxDefaultProps.iconColor}
            size={iconSize as any}
            {...iconProps}
          />
        ))}
    </Box>
  );

  const labelCheckBox = (
    <String
      size={size}
      {...labelProps}
      {...classNameHandler(cn({ [twStyles.isDisabled]: isDisabled }, twStyles.label, labelProps?.className))}
      {...styleHandler({
        defaultStyle: [styles.label, isDisabled && styles.isDisabled],
        style: labelProps?.style,
      })}
    >
      {label}
    </String>
  );

  return (
    <Box>
      <Box {...classNameHandler(twStyles.container)} {...styleHandler({ defaultStyle: styles.container })}>
        <Pressable
          onPress={handlePress}
          disabled={isDisabled}
          {...classNameHandler(twStyles.pressable)}
          {...styleHandler({ defaultStyle: styles.pressable })}
        >
          {contentCheckBox}
          {label && isLabelPressable && labelCheckBox}
        </Pressable>
        {label && !isLabelPressable && labelCheckBox}
      </Box>
    </Box>
  );
}
