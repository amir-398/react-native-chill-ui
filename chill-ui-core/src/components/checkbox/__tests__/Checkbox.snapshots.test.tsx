import renderer, { act } from 'react-test-renderer';
import Checkbox from '../components/Checkbox';

// Mocks
jest.mock('../../../utils', () => ({
  classNamePropsHandler: jest.fn(),
  classNameHandler: jest.fn(() => ({})),
  cn: jest.fn((...args) => args.filter(Boolean).join(' ')),
  isUndefined: jest.fn(value => value === undefined),
  styleHandler: jest.fn(() => ({})),
}));

jest.mock('../../../components/box', () => ({
  Box: ({ children, ...props }: any) => <div {...props}>{children}</div>,
}));

jest.mock('../../../components/icon', () => ({
  Icon: ({ name, ...props }: any) => <div data-testid={`icon-${name}`} {...props} />,
}));

jest.mock('../../../components/string', () => ({
  String: ({ children, ...props }: any) => <div {...props}>{children}</div>,
}));

jest.mock('../styles/Checkbox.ss.styles', () => ({
  checkboxSv: jest.fn(() => ({})),
  styles: { container: {}, contentCheckBox: {}, isDisabled: {}, label: {}, pressable: {} },
}));

jest.mock('../styles/Checkbox.tw.styles', () => ({
  checkboxTv: jest.fn(() => ''),
  twStyles: { container: '', contentCheckBox: '', isDisabled: '', label: '', pressable: '' },
}));

afterEach(() => {
  jest.clearAllTimers();
});

describe('Checkbox Snapshots', () => {
  it('should match snapshot for basic states', () => {
    let tree: any;
    act(() => {
      tree = renderer
        .create(
          <div>
            <Checkbox label="Basic" />
            <Checkbox label="Checked" isChecked={true} />
            <Checkbox label="Disabled" isDisabled={true} />
          </div>,
        )
        .toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for variants', () => {
    let tree: any;
    act(() => {
      tree = renderer
        .create(
          <div>
            <Checkbox label="Circle" variant="circle" />
            <Checkbox label="Square" variant="square" />
          </div>,
        )
        .toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for sizes', () => {
    let tree: any;
    act(() => {
      tree = renderer
        .create(
          <div>
            <Checkbox label="SM" size="sm" />
            <Checkbox label="MD" size="md" />
            <Checkbox label="LG" size="lg" />
          </div>,
        )
        .toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for label behavior', () => {
    let tree: any;
    act(() => {
      tree = renderer
        .create(
          <div>
            <Checkbox label="Pressable" isLabelPressable={true} />
            <Checkbox label="Non-Pressable" isLabelPressable={false} />
          </div>,
        )
        .toJSON();
    });
    expect(tree).toMatchSnapshot();
  });
});
