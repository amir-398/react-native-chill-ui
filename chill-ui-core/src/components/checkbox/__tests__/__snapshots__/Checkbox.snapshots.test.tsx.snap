// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Snapshots should match snapshot for basic states 1`] = `null`;

exports[`Checkbox Snapshots should match snapshot for label behavior 1`] = `null`;

exports[`Checkbox Snapshots should match snapshot for sizes 1`] = `null`;

exports[`Checkbox Snapshots should match snapshot for variants 1`] = `null`;
